Make Subnet Validator Observable

matte21 opened the pull request:
Make Subnet Validator produce Prometheus metrics and K8s API events.
Add Grafana Dashboards and tweak existing ones to reflect the change.
Tweak IPAM Prometheus metrics. Do not apply the same tweaks to the
Connection Agent because it will be heavily refactored soon.

matte21 committed:
Make Subnet Validator Observable

Make Subnet Validator produce Prometheus metrics and K8s API events.
Add Grafana Dashboards and tweak existing ones to reflect the change.
Tweak IPAM Prometheus metrics. Do not apply the same tweaks to the
Connection Agent because it will be heavily refactored soon.

matte21 commented:
Events are emitted every time a subnet status is successfully written, and the "valid subnet" case is differentiated from the "conflicting subnet" case. They are not strictly necessary because they convey nothing which is not in subnet status. I still think we should keep them (I think there are possible use cases, such as a "monitoring controller/system" which can watch only resources of type Event as opposed to monitoring the status of each custom type).

matte21 commented:
The KOS-Internals Grafana Dashboard now has a lot of stuff in it. If you agree, before merging this I would split it into four separate dashboards:

-  KOS Controllers Commons: stuff which exists for each controller and can be printed in the same dashboards; it can be useful to compare the plots of different controllers.
- One dashboard for each separate controller with the stuff that makes sense only for that controller.

There are probably other possible tweaks, but I would add them while we do the performance study.

matte21 commented on staging/kos/metrics/grafana/dashboard-defs/kos-internals.json:
```diff
@@ -142,16 +141,830 @@
+      "steppedLine": false,
+      "targets": [
+        {
+          "expr": "rate(kos_subnet_validator_stale_subnet_work_suppression_count[1m]) / rate(kos_subnet_validator_queue_work_duration_count[1m])",
```
This is not very accurate. At the denominator there's the count of processing runs that were completed, while at the numerator there's the number of times processing stopped because the subnet was stale. Both are relative to the last minute. The inaccuracies are due to the fact that there's no guarantee that counts in the numerator are also counted in the denominator and vice versa. I see no satisfactory to solve the problem, but I think it's ok to keep it as it is, Prometheus is meant to give a coarse-grained idea of what's happening.

MikeSpreitzer replied:
I agree

MikeSpreitzer commented:
If we have a dashboard for each controller then such a dashboard should have all the stuff relevant to that controller, not just the stuff that is not common, in my opinion.
I would leave polishing the dashboards for later, if ever.

MikeSpreitzer commented on staging/kos/pkg/controllers/ipam/ipam.go:
```diff
@@ -128,65 +124,86 @@ func NewController(netIfc kosclientv1a1.NetworkV1alpha1Interface,
+	attachmentUpdateHistograms.With(prometheus.Labels{"statusErr": errValT, "err": errValT})
+	attachmentUpdateHistograms.With(prometheus.Labels{"statusErr": errValF, "err": errValF})
+	attachmentUpdateHistograms.With(prometheus.Labels{"statusErr": errValT, "err": errValF})
+	attachmentUpdateHistograms.With(prometheus.Labels{"statusErr": errValF, "err": errValT})
```
I assume these are added so that scraping produces zeroes rather than nothing?

matte21 replied:
Correct. For histogram quantiles they produce the negative values.

MikeSpreitzer reviewed (commented):
(not done, just pausing)

matte21 commented:
> If we have a dashboard for each controller then such a dashboard should have all the stuff relevant to that controller, not just the stuff that is not common, in my opinion.
> I would leave polishing the dashboards for later, if ever.

Sounds good.

MikeSpreitzer commented on staging/kos/metrics/grafana/dashboard-defs/kos-externals.json:
```diff
@@ -26,11 +25,110 @@
+          "expr": "histogram_quantile(0.99, rate(kos_subnet_validator_subnet_create_to_validated_latency_seconds_bucket{statusErr=\"ok\"}[1m]))",
+          "format": "time_series",
+          "intervalFactor": 2,
+          "legendFormat": "99%ile valid subnets ",
```
I would omit the words that are in the title; make this just "valid".

MikeSpreitzer commented on staging/kos/pkg/controllers/ipam/ipam.go:
```diff
@@ -128,65 +124,86 @@ func NewController(netIfc kosclientv1a1.NetworkV1alpha1Interface,
-	hostname string) *IPAMController {
+	hostname string,
+	metricsNamespace string,
+	metricsSubsystem string) *IPAMController {
```
Only one value appears for `metricsNamespace` here, right?  Only one value appears for `metricsSubsystem` here, right?  This code knows its namespace and subsystem, it does not need to be told by the caller.

MikeSpreitzer commented on staging/kos/pkg/controllers/subnet/validator.go:
```diff
@@ -478,3 +645,11 @@ func equal(s1, s2 []string) bool {
 }
+
+// TODO move to shared util pkg
+func FormatErrVal(err bool) string {
```
Maybe include the abstraction function (`!= nil`) too.

matte21 replied:
See how at https://github.com/MikeSpreitzer/kube-examples/blob/make-subnet-validator-observable/staging/kos/pkg/controllers/subnet/validator.go#L575 that function is used with an argument other than the expression `err != nil`. We can change it to accept an interface argument and check for nilness inside, but then clients which pass a slice in must make sure that it's not empty. I have no preference between keeping things as they are or doing this change.

MikeSpreitzer replied:
Oh, right.  I think it's OK as it is.

MikeSpreitzer commented:
I see a gap in this dashboard.

<img width="1241" alt="Screen Shot 2019-07-22 at 7 33 41 PM" src="https://user-images.githubusercontent.com/14296719/61672307-21f2d380-acb9-11e9-9fec-a2a5b9e506c4.png">

MikeSpreitzer commented:
"Subnet Validator Conflicts Cache Mismatches to Subnets Processed" is the title of a panel showing a ratio.  We also use "to" for time duration.  I recommend a more distinctive indication of a ratio, such as "per" or "/".

MikeSpreitzer commented:
"Subnet Validator Up to date Subnets to Processed Subnets" -> "Fraction of Subnet Updates Not Needed"
(that's `rate(kos_subnet_validator_duplicate_work_count[1m]) / rate(kos_subnet_validator_queue_work_duration_count[1m])`)

"Subnet Validator Useless Live Lists to Live Lists" has related content; can we refactor and put in the same panel?

`rate(kos_subnet_validator_duplicate_work_count[1m]) + sum(rate(kos_subnet_validator_subnet_live_vs_conflict_cache_count[1m])) without (mismatch_type) / ignoring (err) rate(kos_subnet_validator_live_list_latency_seconds_count{err="ok"}[1m])` looks like it is a sum of a rate and a ratio of rates; that seems unlikely to be good.

"Subnet Validator Stale Subnets Suppressions to Processed Subnets" -> "Fraction of Subnet Syncs Ended By Stale Cache"

MikeSpreitzer commented:
"Subnet Validator Worker Utilization" should be adjacent to the related panels, as is done for the other controllers.

MikeSpreitzer commented:
When I go to edit this
<img width="1234" alt="Screen Shot 2019-07-22 at 8 04 19 PM" src="https://user-images.githubusercontent.com/14296719/61673059-063cfc80-acbc-11e9-97aa-efbeda6ede15.png">

I see this:
<img width="1234" alt="Screen Shot 2019-07-22 at 8 04 43 PM" src="https://user-images.githubusercontent.com/14296719/61673070-1228be80-acbc-11e9-9edb-43ae99ae4a5d.png">
(note the shorter panel title).
There seems to be duplication in the original JSON.

matte21 committed:
Tweak Subnet Validator Observability

Do most of the tweaks suggested in PR review. Do other tweaks as well.

matte21 commented:
58628c4 performs most of the fixes and tweaks you suggested:

- Gap in `KOS internals` dashboard is eliminated.
- IPAM and Subnet validator have their metrics namespace and subsystem as internal consts as opposed to being told by the caller. 
- Some metrics legends are shortened to avoid redundancy with the panel title.
- Some panel titles are improved.
- Ordering of panels is improved to keep relevant ones close.
- Merge panels with related content into a single one.

Concerning the last point, you wrote:

> "Subnet Validator Up to date Subnets to Processed Subnets" -> "Fraction of Subnet Updates Not Needed"
> (that's rate(kos_subnet_validator_duplicate_work_count[1m]) / rate(kos_subnet_validator_queue_work_duration_count[1m]))
> 
> "Subnet Validator Useless Live Lists to Live Lists" has related content; can we refactor and put in the same panel?

The two panels have been merged into a single one together with the one about the number of mismatches between live list results and conflicts cache owner. The new panel shows the fraction of useless live lists, both on aggregate and per-single cause of uselessness (subnet status is up to date, missing conflict cache, conflict cache owner, conflict cache mismatch in general). Note that for consistency this required changing the metrics of the panels about up to date subnets and conflicts cache mismatches: in both cases the denominator went from `rate(kos_subnet_validator_queue_work_duration_count[1m])` to `rate(kos_subnet_validator_live_list_latency_seconds_count{err="ok"}[1m])`.

Also, you wrote:
> rate(kos_subnet_validator_duplicate_work_count[1m]) + sum(rate(kos_subnet_validator_subnet_live_vs_conflict_cache_count[1m])) without (mismatch_type) / ignoring (err) rate(kos_subnet_validator_live_list_latency_seconds_count{err="ok"}[1m]) looks like it is a sum of a rate and a ratio of rates; that seems unlikely to be good.

Good catch. Everything to the left of `/`  was meant to be wrapped by `()`.

matte21 commented:
I would make one last tweak if you agree: merging the panels about worker utilization and worker utilization for calls out.

matte21 committed:
Update Grafana Dashboards with Queue Metrics

Update Grafana dashboards using K8s workqueue Prometheus metrics to use
the new metrics, whose unit of measure is seconds, as opposed to the
deprecated ones which use microseconds.

matte21 commented:
I realized that Grafana panels using metrics about workqueues still used the deprecated metrics in https://github.com/MikeSpreitzer/kube-examples/blob/make-subnet-validator-observable/staging/kos/pkg/controllers/workqueue_prometheus/connect.go based on microseconds. c261ded switches to the new metrics based on seconds.

MikeSpreitzer commented on staging/kos/pkg/controllers/workqueue_prometheus/connect.go:
```diff
@@ -30,8 +30,9 @@ import (
 // prometheus metrics. To use this package, you just have to import it.
 
-// Metrics subsystem and keys used by the workqueue.
+// Metrics namespace, subsystem and keys used by the workqueue.
```
Ugh, I do not like not having an exact copy.  OTOH, the upstream source oddly lacks the Prometheus metrics namespace.  I guess there is an issue to raise upstream before we can do better here.

MikeSpreitzer commented:
> I would make one last tweak if you agree: merging the panels about worker utilization and worker utilization for calls out.

That would be OK for the centralized controllers.  For the connection agent there is also worker utilization for calls to the fabric, and a lot of traces; I would do no merging for those.

MikeSpreitzer commented:
For the connection agent Grafana panels:

- the panel showing worker utilization summed over workers should, but no longer does, explicitly say it is summed over workers.

- for status updates there should be, as for other controllers that make apiserver calls:

- - utilization summed over workers (this is missing);
- - 99th percentile of latency;
- - average latency;
- - rate (this is missing).

The status update panels should all precede or all follow all the fabric panels.

MikeSpreitzer commented:
Issues noted above can be addressed in follow-up PRs.
